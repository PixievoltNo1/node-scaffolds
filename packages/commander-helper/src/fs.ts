import { isFileSync } from '@guanghechen/file-helper'
import fs from 'fs-extra'
import yaml from 'js-yaml'
import path from 'path'

/**
 * Load configuration file with format .json / .yml / .yaml
 *
 * @param filepath
 * @param logger
 */
export async function loadJsonOrYaml(
  filepath: string,
  encoding = 'utf-8',
): Promise<unknown | never> {
  if (!isFileSync(filepath))
    throw new Error(`${filepath} is an invalid file path`)

  const loadContent = (): Promise<string> =>
    fs.readFile(filepath, encoding as BufferEncoding)

  let result: unknown
  const extname = path.extname(filepath)
  switch (extname) {
    case '.json':
      {
        const content: string = await loadContent()
        result = JSON.parse(content)
      }
      break
    case '.yml':
    case '.yaml':
      {
        const content: string = await loadContent()
        result = yaml.load(content, { filename: filepath, json: true })
      }
      break
    default:
      throw new Error(
        `Only files in .json / .yml / .ymal format are supported. filepath(${filepath}`,
      )
  }
  return result
}

/**
 * Load configuration file with format .json / .yml / .yaml  (synchronizing)
 *
 * @param filepath
 * @param logger
 */
export function loadJsonOrYamlSync(
  filepath: string,
  encoding = 'utf-8',
): unknown | never {
  if (!isFileSync(filepath))
    throw new Error(`${filepath} is an invalid file path`)

  const loadContent = (): string =>
    fs.readFileSync(filepath, encoding as BufferEncoding)

  let result: unknown
  const extname = path.extname(filepath)
  switch (extname) {
    case '.json':
      {
        const content: string = loadContent()
        result = JSON.parse(content)
      }
      break
    case '.yml':
    case '.yaml':
      {
        const content: string = loadContent()
        result = yaml.load(content, { filename: filepath, json: true })
      }
      break
    default:
      throw new Error(
        `Only files in .json / .yml / .ymal format are supported. filepath(${filepath}`,
      )
  }
  return result
}
