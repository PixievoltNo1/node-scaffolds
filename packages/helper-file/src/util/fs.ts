import fs from 'node:fs'
import path from 'node:path'

/**
 * Collect all files under the given directory.
 *
 * @param dir
 * @param predicate
 */
export async function collectAllFiles(
  dir: string,
  predicate?: ((p: string, stat: fs.Stats) => Promise<boolean> | boolean) | null,
): Promise<string[]> {
  const results: string[] = []
  await collect(dir)
  return results

  async function collect(filepath: string): Promise<void> {
    const stat = fs.statSync(filepath)
    if (stat.isDirectory()) {
      const filenames = fs.readdirSync(filepath)
      for (const filename of filenames) {
        const nextFilepath = path.join(filepath, filename)
        await collect(nextFilepath)
      }
      return
    }

    if (stat.isFile()) {
      if (predicate == null || (await predicate(filepath, stat))) {
        results.push(filepath)
      }
    }
  }
}

/**
 * Collect all files under the given directory.   (synchronizing)
 *
 * @param dir
 * @param predicate
 */
export function collectAllFilesSync(
  dir: string,
  predicate?: ((p: string, stat: fs.Stats) => boolean) | null,
): string[] {
  const results: string[] = []
  collect(dir)
  return results

  function collect(filepath: string): void {
    const stat = fs.statSync(filepath)
    if (stat.isDirectory()) {
      const filenames = fs.readdirSync(filepath)
      for (const filename of filenames) {
        const nextFilepath = path.join(filepath, filename)
        collect(nextFilepath)
      }
      return
    }
    if (stat.isFile()) {
      if (predicate == null || predicate(filepath, stat)) {
        results.push(filepath)
      }
    }
  }
}
