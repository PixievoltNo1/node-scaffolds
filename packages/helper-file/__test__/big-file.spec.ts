import { desensitize, locateFixtures, unlinkSync } from 'jest.helper'
import { existsSync, readFileSync, statSync } from 'node:fs'
import fs from 'node:fs/promises'
import {
  BigFileHelper,
  bigFileHelper,
  calcFilePartItemsByCount,
  calcFilePartItemsBySize,
} from '../src'

describe('calcFilePartItemsByCount', () => {
  test('basic', async () => {
    const filepath = locateFixtures('a.md')
    expect(
      calcFilePartItemsByCount(await fs.stat(filepath).then(md => md.size), 4),
    ).toMatchSnapshot()
  })

  test('empty file', async () => {
    const filepath = locateFixtures('b.md')
    expect(calcFilePartItemsByCount(await fs.stat(filepath).then(md => md.size), 1024)).toEqual([
      { sid: 1, start: 0, end: 0 },
    ])
  })
})

describe('calcFilePartItemsBySize', () => {
  test('basic', async () => {
    const filepath = locateFixtures('a.md')
    expect(
      calcFilePartItemsBySize(await fs.stat(filepath).then(md => md.size), 1024),
    ).toMatchSnapshot()
  })

  test('empty file', async () => {
    const filepath = locateFixtures('b.md')
    expect(calcFilePartItemsBySize(await fs.stat(filepath).then(md => md.size), 1024)).toEqual([
      { sid: 1, start: 0, end: 0 },
    ])
  })
})

describe('BigFileHelper', () => {
  const filepath = locateFixtures('a.md')
  const outputFilepath = filepath + Math.random()
  const parts = calcFilePartItemsByCount(statSync(filepath).size, 5)
  const partFilepaths = bigFileHelper.calcPartFilepaths(filepath, parts)
  const originalContent = readFileSync(filepath, bigFileHelper.encoding)

  afterEach(() => {
    unlinkSync(outputFilepath, partFilepaths)
  })

  describe('calcPartFilepaths', () => {
    test('basic', () => {
      expect(desensitize(partFilepaths)).toMatchSnapshot()
      expect(bigFileHelper.calcPartFilepaths(filepath, [])).toEqual([filepath])
      expect(bigFileHelper.calcPartFilepaths(filepath, [{} as any])).toEqual([filepath])
    })

    test('custom part name', () => {
      const fileHelper = new BigFileHelper({ partSuffix: '.fun' })
      const funPartFilepaths = fileHelper.calcPartFilepaths(filepath, parts)
      expect(desensitize(funPartFilepaths)).toMatchSnapshot()
    })
  })

  test('split', async () => {
    for (const filepath of partFilepaths) {
      expect(existsSync(filepath)).toBe(false)
    }

    expect(await bigFileHelper.split(filepath, [])).toEqual([filepath])
    for (const filepath of partFilepaths) {
      expect(existsSync(filepath)).toBe(false)
    }

    expect(await bigFileHelper.split(filepath, parts)).toEqual(partFilepaths)

    for (const filepath of partFilepaths) {
      expect(existsSync(filepath)).toBe(true)
    }
  })

  test('merge', async () => {
    for (const filepath of partFilepaths) {
      expect(existsSync(filepath)).toBe(false)
    }

    expect(existsSync(outputFilepath)).toBe(false)
    expect(await bigFileHelper.split(filepath, parts)).toEqual(partFilepaths)

    for (const filepath of partFilepaths) {
      expect(existsSync(filepath)).toBe(true)
    }

    await bigFileHelper.merge(partFilepaths, outputFilepath)
    expect(existsSync(outputFilepath)).toBe(true)
    expect(readFileSync(outputFilepath, bigFileHelper.encoding)).toEqual(originalContent)
  })
})
