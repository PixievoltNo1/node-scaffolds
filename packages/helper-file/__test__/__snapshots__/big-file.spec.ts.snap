// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BigFileHelper calcPartFilepaths basic 1`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.ghc-part1",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.ghc-part2",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.ghc-part3",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.ghc-part4",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.ghc-part5",
]
`;

exports[`BigFileHelper calcPartFilepaths custom part name 1`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.fun1",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.fun2",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.fun3",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.fun4",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/a.md.fun5",
]
`;

exports[`calcFilePartItemsByCount basic 1`] = `
Array [
  Object {
    "end": 1336,
    "sid": 1,
    "start": 0,
  },
  Object {
    "end": 2672,
    "sid": 2,
    "start": 1336,
  },
  Object {
    "end": 4008,
    "sid": 3,
    "start": 2672,
  },
  Object {
    "end": 5341,
    "sid": 4,
    "start": 4008,
  },
]
`;

exports[`calcFilePartItemsBySize basic 1`] = `
Array [
  Object {
    "end": 1024,
    "sid": 1,
    "start": 0,
  },
  Object {
    "end": 2048,
    "sid": 2,
    "start": 1024,
  },
  Object {
    "end": 3072,
    "sid": 3,
    "start": 2048,
  },
  Object {
    "end": 4096,
    "sid": 4,
    "start": 3072,
  },
  Object {
    "end": 5120,
    "sid": 5,
    "start": 4096,
  },
  Object {
    "end": 5341,
    "sid": 6,
    "start": 5120,
  },
]
`;
